core/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -648,12 +652,60 @@ private void checkPartitionedIndexConstraints(DocumentMapper newMapper) {
         }
     }
 
-    private void checkIndexSortCompatibility(IndexSortConfig sortConfig, boolean hasNested) {
+    private static void checkIndexSortCompatibility(IndexSortConfig sortConfig, boolean hasNested) {
         if (sortConfig.hasIndexSort() && hasNested) {
             throw new IllegalArgumentException("cannot have nested fields when index sort is activated");
         }
     }
 
+    private static void checkCopyToValidity(List<FieldMapper> fieldMappers, Map<String, ObjectMapper> fullPathObjectMappers) {
+        for (FieldMapper mapper : fieldMappers) {
+            if (mapper.copyTo() != null && mapper.copyTo().copyToFields().isEmpty() == false) {
+                final String sourceScope = getNestedScope(mapper.name(), fullPathObjectMappers);
+                for (String copyTo : mapper.copyTo().copyToFields()) {
+                    if (fullPathObjectMappers.containsKey(copyTo)) {
+                        throw new IllegalArgumentException("Cannot copy to field [" + copyTo + "] since it is mapped as an object");
+                    }
+                    final String targetScope = getNestedScope(copyTo, fullPathObjectMappers);
+                    checkNestedScopeCompatibility(sourceScope, targetScope);
```

Comment:
I like having `checkNestedScopeCompatibility` entirely responsible for checking what is allowed and what is not.